### PR TITLE
Show message box default button

### LINKED ISSUE #3810: dialog.showMessageBox() should have an option to set the primary button
I am trying to produce a correct dialog on Linux where typically you have "Alternate" | "Cancel" | "Ok" as button order. However when I pass in like that, "Alternate" is being focussed by default.

I think it would be good to have API to set the primary button on the dialog so that I can make "Ok" the default

### PR DESCRIPTION
**Pull request by leethomas**
Closes #3810.

~~Only supported & tested on OS X & Linux, as I don't have a Windows machine. I updated the `showMessageBox` method signatures for Windows to accept the extra parameter so it (hopefully) wouldn't break for Windows users. Please test on Windows before merging.~~ Now supported on Windows, thanks to etikin

Example:
`dialog.showMessageBox(win, { buttons: ['hi', 'there', 'cheng', 'zhao'], message: 'hi there!', title: 'testing', defaultButtonIndex: 2, type: 'info'});`

OS X
screen shot 2016-01-07 at 10 52 58 pm
Just fyi, OS X displays buttons from right to left, starting with the first button (0 index) at the furthest right. So the above is expected behavior.

Linux
usee

Thanks!

### CODE REVIEW

**Comment by leethomas** [comment]
fixing lint/code format errors

**Comment by leethomas** [comment]
Fixed Windows compilation errors

File: atom/browser/api/lib/dialog.coffee [hunk 1]
```diff
@@ -108,6 +109,7 @@ module.exports =
 
     binding.showMessageBox messageBoxType,
                            options.buttons,
+                           options.defaultButtonIndex,
```
Comment (etiktin) [inline atom/browser/api/lib/dialog.coffee]: I think the new option should be called `defaultId` to match the way we call the cancel button index (`cancelId`).

**Comment by etiktin** [comment]
You can add support to Windows, by setting the `nDefaultButton` key of the `TASKDIALOGCONFIG` struct. So if a default was provided you need to set it to `kIDStart + default_button_index`.

File: atom/browser/api/lib/dialog.coffee (hunk 1 above)
Comment (leethomas) [inline atom/browser/api/lib/dialog.coffee]: Makes sense, will do

**Comment by leethomas** [comment]
@etiktin Done, thank you! Please test when you get a chance. I assume you're using Windows.

File: atom/browser/api/lib/dialog.coffee [hunk 2]
```diff
@@ -91,10 +91,11 @@ module.exports =
...
 
-    options.title ?= ''
-    options.message ?= ''
-    options.detail ?= ''
-    options.icon ?= null
+    options.title              ?= ''
+    options.message            ?= ''
+    options.detail             ?= ''
+    options.icon               ?= null
+    options.defaultId ?= -1
```
Comment (etiktin) [inline atom/browser/api/lib/dialog.coffee]: The spacing in lines 94-98 is not consistent. There should be a single space before `?=`.

File: atom/browser/ui/message_box_mac.mm [hunk 3]
```diff
@@ -78,8 +79,17 @@ - (void)alertDidEnd:(NSAlert*)alert
...
     if (buttons[i].empty())
       title = @"(empty)";
+
     NSButton* button = [alert addButtonWithTitle:title];
     [button setTag:i];
+
+    if (i == (size_t)default_id) {
+      // focus the button at default_id if the user opted to do so.
+      // The first button added gets set as the default selected.
+      // So remove that default, and make the requested button the default
```
Comment (etiktin) [inline atom/browser/ui/message_box_mac.mm]: We usually start a comment with a capital letter and end with a period.

File: docs/api/dialog.md [hunk 4]
```diff
@@ -87,6 +87,7 @@ will be passed via `callback(filename)`
   `"warning"`. On Windows, "question" displays the same icon as "info", unless
   you set an icon using the "icon" option.
   * `buttons` Array - Array of texts for buttons.
+  * `defaultId` Integer - Index of the button in the buttons array which will be selected by default when the message box opens.
```
Comment (etiktin) [inline docs/api/dialog.md]: Try to limit line length to 80 characters.

**Comment by etiktin** [comment]
I tested it on Windows and it works!
Added some stylstic comments, but other than that it LGTM :+1:

**Comment by zcbenz** [comment]
Thanks!

**Comment by bpasero** [comment]
Thanks <3

**Pull request merged**